src/classes/Accounts.cls

```diff
@@ -0,0 +1,154 @@
+        this.oldMap = new Map<Id, Account>(oldlist);
+    }
+
+    public void onBeforeInsert() {
+        setAccountTypeAndIndividualFieldValues();
+    }
+
+    public void onBeforeUpdate() {
+        setAccountTypeAndIndividualFieldValues();
+        setSystemFieldValues();
+        buildHHLists();
+        if (listHHNew.size() > 0 && service.isAdvancedHouseholdNaming()) {
+            if (!isHouseholdTriggerContext()) {
+                service.setHouseholdNameFields(listHHNew, oldMap);
+            }
+        }
+    }
+
+    public void onAfterUpdate() {
+        buildHHLists();
+        if (listHHNew.size() > 0 && service.isAdvancedHouseholdNaming()) {
+            List<Id> householdsNeedingNameUpdates =
+                    service.getHouseholdsNeedingNameUpdates(
+                            listHHNew, oldMap
+                    );
+            if (!householdsNeedingNameUpdates.isEmpty()) {
+                service.renameHouseholdAccounts(householdsNeedingNameUpdates);
+            }
+        }
+    }
+
+    public void setSystemFieldValues() {
+        Integer i = 0;
+        for (Object obj : getRecords()) {
+            Account acc = (Account) obj;
+
+            if (!isHouseholdAccount(acc)
+                    && !isOneToOneAccount(acc) && !isBucketAccount(acc)) {
+                if (isRecordTypeIdChanged(acc, i)) {
+                    acc.npe01__SYSTEMISINDIVIDUAL__c = false;
+                    acc.npe01__SYSTEM_AccountType__c = null;
+                }
+            }
+            i++;
+        }
+    }
+
+    private Boolean isRecordTypeIdChanged(Account acc, Integer i) {
+        return acc.RecordTypeId != listOld[i].get('RecordTypeId');
+    }
+
+    public void setAccountTypeAndIndividualFieldValues() {
+        integer i = 0;
+        for (SObject so : getRecords()) {
+            Account acc = (Account) so;
+
+            if (isHouseholdAccount(acc)) {
+                if (acc.npe01__SYSTEM_AccountType__c == null || acc.npe01__SYSTEMISINDIVIDUAL__c != true) {
+                    acc.npe01__SYSTEM_AccountType__c = CAO_Constants.HH_ACCOUNT_TYPE;
+                    acc.Type = CAO_Constants.HH_TYPE;
+                    acc.npe01__SYSTEMISINDIVIDUAL__c = true;
+                }
+            }
+
+            else if (isOneToOneAccount(acc)) {
+                if (acc.npe01__SYSTEM_AccountType__c == null || acc.npe01__SYSTEMISINDIVIDUAL__c != true) {
+                    acc.npe01__SYSTEM_AccountType__c = CAO_Constants.ONE_TO_ONE_ORGANIZATION_TYPE;
+                    acc.npe01__SYSTEMISINDIVIDUAL__c = true;
+                }
+            }
+
+            i++;
+        }
+    }
+
+    private void buildHHLists() {
+        integer i = 0;
+        for (SObject so : getRecords()) {
+            Account acc = (Account) so;
+            if (acc.npe01__SYSTEM_AccountType__c == CAO_Constants.HH_ACCOUNT_TYPE) {
+                listHHNew.add(acc);
+                listHHOld.add((Account) listOld[i]);
```

Comment:
Let's use the oldMap instead of the index of oldList? We can't rely on the order in Trigger.old and Trigger.new.